### PR TITLE
[DD4hep] Ecal Barrel algorithm - use assemblies. Fix all overlaps in EBAR.

### PR DESCRIPTION
**Pull request by ianna**
#### PR description:

Two `SubtractionSolids` volumes: `ESPM` and `EHAWR` made of `E_Air` are replaced by `Assemblies` with identical names. One `SubtractionSolid` `EFAW` is left as a mother volume intentionally in the hierarchy. This helps to restrict the number of volumes for a ROOT overlap check and complete it within a reasonable time.
 
Screenshot 2021-04-06 at 13 18 57

The main source of 36 overlaps for both DD and DD4hep-based EcalBarrel geometry was a rounding error. This is solved by adding 1 mm to a Z in `EBAR` envelope (in blue):

Screenshot 2021-04-06 at 13 54 59
After this change there are no overlaps reported in Ecal barrel in DD4hep-based geometry:
Screenshot 2021-04-06 at 14 43 12

@bsunanda - FYI

#### if this PR is a backport please specify the original PR and why you need to backport that PR:

Before submitting your pull requests, make sure you followed this checklist:
- verify that the PR is really intended for the chosen branch
- verify that changes follow [CMS Naming, Coding, And Style Rules](http://cms-sw.github.io/cms_coding_rules.html)
- verify that the PR passes the basic test procedure suggested in the [CMSSW PR instructions](https://cms-sw.github.io/PRWorkflow.html)

### CODE REVIEW

**Comment by ianna** [comment]
please test

File: Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc [hunk 1]
```diff
@@ -2087,7 +2048,7 @@ static long algorithm(dd4hep::Detector& /* description */, cms::DDParsingContext
       Volume backPlateLog = Volume(back.plateName, backPlateSolid, ns.material(back.plateMat));
 
       const Position backPlateTra(
-          0.5 * back.sideHeight + backPlateParms[1], 0, backPlateParms[2] - 0.5 * back.sideLength);
+          0.5 * back.sideHeight + backPlateParms[1], 0 * dd4hep::mm, backPlateParms[2] - 0.5 * back.sideLength);
```
Comment (civanch) [inline Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc:2051]: @ianna , here 0 is more correct then 0*mm

File: Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc [hunk 2]
```diff
@@ -2144,27 +2106,29 @@ static long algorithm(dd4hep::Detector& /* description */, cms::DDParsingContext
       Solid backSideSolid = mytrap(back.sideName, trapBS);
       Volume backSideLog = Volume(back.sideName, backSideSolid, ns.material(back.sideMat));
 
-      const Position backSideTra1(0, back.plateWidth / 2 + back.sideYOff1, 1 * dd4hep::mm);
+      const Position backSideTra1(0 * dd4hep::mm, back.plateWidth / 2. + back.sideYOff1, 1 * dd4hep::mm);
```
Comment (civanch) [inline Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc:2109]: Here I would use "0" as well, if it is changed from "/2" to "/2." I would use "*0.5", unfortunately the effect of multiplication may be seen if it is applied everywhere, not in a single line.

**Comment by cvuosalo** [comment]
Clang warnings need to be fixed:

Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-04-05-2300/src/Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc:965:19: warning: unused variable 'alltrot' [-Wunused-variable]
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-04-05-2300/src/Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc:950:11: warning: unused variable 'temp2' [-Wunused-variable]
/data/cmsbld/jenkins/workspace/ib-run-pr-tests/CMSSW_11_3_X_2021-04-05-2300/src/Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc:949:11: warning: unused variable 'temp1' [-Wunused-variable]

File: Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc [hunk 3]
```diff
@@ -1231,10 +1194,12 @@ static long algorithm(dd4hep::Detector& /* description */, cms::DDParsingContext
...
               ++fanOutCopy;
-              xilyLog.placeVolume(ilyFanOutLog,
-                                  fanOutCopy,
-                                  Transform3D(RotationZ(phi) * RotationY(180_deg),
-                                              Position(xx, yy, ily.vecIlyFanOutZ[ilyFO] - ilyLengthHalf)));
+              xilyLog.placeVolume(
+                  ilyFanOutLog,
+                  fanOutCopy,
+                  Transform3D(
+                      RotationZ(phi) * RotationY(180_deg),  // CLHEP::HepRotationZ(phi) * CLHEP::HepRotationY(180 * deg)
```
Comment (cvuosalo) [inline Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc:1201]: Why is the CLHEP comment being added to this line?

Comment (ianna) [reply Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc:1201]: removed

File: Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc (hunk 1 above)
Comment (ianna) [reply Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc:2051]: @civanch - the effect is the same. IMHO, adding units to a number explains better what the number is. There is no downside - the compiler will optimise it anyway. :-)

File: Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc (hunk 2 above)
Comment (ianna) [reply Geometry/EcalCommonData/plugins/dd4hep/DDEcalBarrelNewAlgo.cc:2109]: @civanch - I've changed the division by the multiplication. Thanks!

**Comment by ianna** [comment]
please test

**Comment by ianna** [comment]
please test

**Comment by cvuosalo** [comment]
@cmsbuild please test workflow 11634.911, 11650.911

**Comment by civanch** [comment]
Minor change in ECAL geometry provoke change in simulation history, so no regression in this PR.

**Comment by cvuosalo** [comment]
This PR still has a problem with the ECal barrel:

%MSG-w EcalGeom: OscarMTProducer:g4SimHits 07-Apr-2021 21:16:39 CEST Run: 1 Event: 1
ECalBarrelNumberingScheme::getUnitID(): Not enough levels found in EcalBaseNumber ( 11) Returning 0
%MSG

There are thousands of these messages in the workflow logs.

**Comment by silviodonato** [comment]
cc @cms-sw/ecal-dpg-l2

**Comment by bsunanda** [comment]
@ianna @silviodonato @qliphy This PR needs the new dd4hep version to work. The new dd4hep is indeed needed

**Comment by civanch** [comment]
@ianna , @bsunanda , @mrodozov , if special IB made recently by Mirco is working why not to make a new version of DD4Hep for 12_0_X? Or DD4Hep update is not yet tagged?

**Comment by civanch** [comment]
The issue https://github.com/cms-sw/cmsdist/issues/6824 is created.

**Comment by ianna** [comment]
@civanch - I guess this one can go in now. It will need a PR from @bsunanda to update the numbering.

**Comment by bsunanda** [comment]
@ianna @civanch @silviodonato The PR for numbering is already integrated with the master. So with the new version of dd4hep, this PR can be integrated.

**Comment by civanch** [comment]
+1

**Comment by silviodonato** [comment]
+1

**Pull request merged**